### ISSUE TITLE
[Histogram] odd colors

### ISSUE THREAD

**Issue by AxelG-DE** [darktable-org/darktable#6945]
**Describe the bug**

In my histogram I have odd colors, but I admit, it might be intentional

**Expected behavior**

as previously

**Screenshots**

NEW (no groups is due to bisecting)
![Screenshot_20201116_150943](https://user-images.githubusercontent.com/49004103/99265215-bb84b380-2821-11eb-8de7-00bf917c793f.png)

previous
![Screenshot_20201116_151157](https://user-images.githubusercontent.com/49004103/99264594-f4705880-2820-11eb-9707-605d383e345a.png)

**Bisect**
bisect gave me:

```
855c230401f0c585b38c3ab7660d37f9d034f13a is the first bad commit
commit 855c230401f0c585b38c3ab7660d37f9d034f13a
Author: Dan Torop <dant@pnym.net>
Date:   Sat Nov 14 11:57:24 2020 -0500

    histogram: draw waveform in display profile (#6135)
    
    * histogram: draw waveform in display profile
    
    Generate waveform and RGB parade in Linear Rec709 RGB. Convert to
    display colorspace before drawing.
    
    Advantages:
    
    - Gets rid of gamma code when processing image to produce a
    waveform. This lets the process happen extremely quickly, in a single
    loop. No longer need code to cache powf() output to speed up this
    transformation. Use darktable's fast colorspace conversion code instead.
    
    - In a wide gamut display, if blue is rendered as (0,0,255) in display
    profile, it may practically invisible. (0,0,255) in Rec.709 is more
    visible.
    
    - The new code allocates 3 times the amount of permanent memory (now
    360 x 175 x (sizeof(float) x 2 + sizeof(uint8_t))) = 553K, vs. 184K in
    the old code. But the old code would allocate approx. 379K to process
    the histogram and 984K to draw a waveform histogram. The new code
    doesn't use any such temporary buffers. There's no malloc needed during
    processing or drawing.
    
    * histogram: use stylesheet to set waveform colors
    
    Set specific values for red, green, and blue in the waveform, rather
    than using the primaries of Rec.709. Switch to drawing in Rec.2020 --
    a reasonable colorspace for contemporary monitors.
    
    * histogram: use stylesheet colors for RGB parade
    
    * histogram: use stylesheet colors for regular histogram

 data/themes/darktable.css |   1 +
 src/bauhaus/bauhaus.c     |   1 +
 src/bauhaus/bauhaus.h     |   2 +-
 src/iop/rgbcurve.c        |  12 +-
 src/iop/rgblevels.c       |  12 +-
 src/libs/histogram.c      | 370 ++++++++++++++++++++++++++++------------------
 src/libs/lib.h            |   4 +-
 7 files changed, 244 insertions(+), 158 deletions(-)
```
@dtorop may I ping you here?
I do have wide gammut displays, both are calibrated both show a bit different result, while the to be edited picture is always the same 

[Edit1] I checked on my Lenovo T450s, also have this a bit odd colors (no wide gammut display, but IPS)

**Platform (please complete the following information):**
* darktable version : 3.3.0+1649~g1545a4ba2
* OS                : Linux - kernel 5.9.8-gentoo
* Distro            : Gentoo Base System release 2.7
* Processor         : Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
* Memory            : 32 GB (4 x 8 GB) + 5GB Swap
* Graphics card0    : GeForce GTX 1060 6GB
* Graphics card1    : GeForce RTX 2070 SUPER
* Graphics driver   : nvidia-drivers-455.38
* OpenCL installed  : Yes (opencl-headers-2020.06.16)
* OpenCL activated  : Yes
* Xorg              : xorg-server-1.20.8-r1
* Desktop           : KDE 5
* GTK+              : gtk+-3.24.22
* gcc               : x86_64-pc-linux-gnu-9.3.0
* cflags            : CFLAGS="-march=native-O2-mtune=native-pipe"
* CMAKE_BUILD_TYPE  : "Release"

**Comment by johnny-bit** [comment]
Yea, this is due to @dtorop changes in histogram primaries...

**Comment by Nilvus** [comment]
Commit you point were the first release of those colors (more "odd" ones that what you point there). @dtorop have improved them after my feedback in that PR: #6922. You have colors improved in that last PR and yes they are less saturated than before. Before, they were just too saturated and we were used to. With new colors, transitions are better and more flashy. Histogram is just an help to process images, and so should not be the main part. Flashy/over-saturated colors like before could have drawback. You could understand more why that choice by reading first PR #6135 and PR #6922 that improve colors or simply last comments by @dtorop and my answer in PR #6922.

**Comment by Nilvus** [comment]
Last thing: seeing your screenshots, I just found colors more suitable in new way than on old way. It's a new habit to have and you will soon forgot last colors!

**Comment by kofa73** [comment]
My histogram is pinkish, instead of white. I don't observe a similar colour shift anywhere else on the UI.
![Screenshot_20201116_184845](https://user-images.githubusercontent.com/3177739/99311718-48009780-285d-11eb-9529-50720ad2da6e.png)

**Comment by AxelG-DE** [comment]
> Last thing: seeing your screenshots, I just found colors more suitable in new way than on old way. It's a new habit to have and you will soon forgot last colors!

I will read those PRs, you mentioned above.
Seriously, atm I can hardly imagine, I will like it. White has a too pinky touch and yellow is not yellow anymore but chester-cheese-color. Not a clean  straight forward color, hence atm not really appreaceated. Sorry, being so frankly.

**Comment by Nilvus** [comment]
> My histogram is pinkish, instead of white. I don't observe a similar colour shift anywhere else on the UI. The image here is a grey ramp generated in the Gimp (the top half has a LUT applied, as the screenshot was originally not made for this issue; the bottom half is the input).

You don't precise which darktable version/commit you use and it's not the same issue (even if it's a histogram color). We talk here about colors that are less saturated that before not strange colors in histogram like your pink one. I suggest you to create another issue and post also the file to let a developer test it.

**Comment by Nilvus** [comment]
> Seriously, atm I can hardly imagine, I will like it. White has a too pinky touch and yellow is not yellow anymore but chester-cheese-color. Not a clean straight forward color, hence atm not really appreaceated. Sorry, being so frankly.

No problem, that was also my first way to see those new colors. And indeed it's because we were too used to over-saturated colors in histogram in darktable. At first, I dislike them but with last improved colors and some time, I had understood that they really were over-saturated before. I suggest you to find and check some histogram screen captures of other software for example, you will see that no histogram has a real white and you will be surprise to some histograms like for example Rawtherapee, ARTherapee, Lightroom, etc.

But new colors could maybe be improved again. I'm ok with those new colors as I like also old ones. What I like most with new histogram colors is that they are a lot less distracting than old colors. And that is a good thing as it's better to process an image with our eyes than with the histogram (the histogram is just here as a help).

**Comment by AxelG-DE** [comment]
My dt-version is listed in the issue.
I really don't see why I should release another issue. This **IS** the issue about **odd colours** in the **histogram** and NOT about more or less saturation, and nothing else. Just you turned it into this direction. 

Further, there is no file to be submitted, it is all about histogram only.

BTW: this is on my EIZO monitor. When I push dt over to my HP monitor, at least the odd pinky in the white is (partially) gone, but the rest remains the same (and again, on my Lenovo it is also oddy pinkish)

**Comment by kofa73** [comment]
> > My histogram is pinkish, instead of white. I don't observe a similar colour shift anywhere else on the UI. The image here is a grey ramp generated in the Gimp (the top half has a LUT applied, as the screenshot was originally not made for this issue; the bottom half is the input).
> 
> You don't precise which darktable version/commit you use

OS: Kubuntu 20.10
darktable-cmstest version 3.3.0+1655~gecd7e2cbf
this executable was built with colord support enabled
darktable itself was built with colord support enabled

primary CRTC is at CRTC 0
CRTC for screen 0 CRTC 1 has no mode or no output, skipping
CRTC for screen 0 CRTC 2 has no mode or no output, skipping
CRTC for screen 0 CRTC 3 has no mode or no output, skipping

DP-2    the X atom and colord returned the same profile
        X atom: _ICC_PROFILE (21928 bytes)
                description: HP ZR22w 2015-10-11 D6500 2.2 M-S 3xCurve+MTX
        colord: "/var/lib/colord/icc/HP ZR22w 2015-10-11 D6500 2.2 M-S 3xCurve+MTX.icc"
                description: HP ZR22w 2015-10-11 D6500 2.2 M-S 3xCurve+MTX

Your system seems to be correctly configured


 and it's not the same issue (even if it's a histogram color). We talk here about colors that are less saturated that before not strange colors in histogram like your pink one. I suggest you to create another issue and post also the file to let a developer test it.

**Comment by Nilvus** [comment]
@kofa73: your answer is no help. Read again my answer to you.

**Comment by dtorop** [comment]
Interesting reports/dialog. Unfortunately, I can absolutely believe that the white in the histogram has turned pink in some displays/profiles.

Here's a line of thought: The primary colors, post PR #6135 are specified as Lab values, which are translated into display RGB. I chose Lab values (in both #6135 and #6922) such that they would produce a neutral gray (red, green, and blue channels the same in display RGB) when overlaid. My hope was that the magic of chromatic adaptation would make for a neutral gray regardless of the illuminant of the display profile. This may have been wishful thinking.

Maybe another way to handle this: Specify the neutral gray for when all three channels overlay in the histogram. Specify the lightness and hue of each primary. Then calculate primaries such that each has the same chromaticity distance from neutral gray. The primaries should then overlay to produce neutral gray. If the chroma of the primaries is well chosen, the secondaries should also be reasonable.

This may be a whole can of worms, but the math would only need to be done once for a given display profile. It's important to fix if the histogram white is not neutral for certain displays/profiles. Besides being un-aesthetic, having any big area of non-neutral color near the image could make color balancing more error-prone.

This all is totally different from just plain not liking the choice of primaries -- though I can sympathize with that too.

Looking at the sample from @AxelG-DE, I see "before" neutral as (222,222,222), "after" as (241,233,245) which would certainly produce an annoying color cast. The image from @kofa73 the "neutral" is worse (226,204,252). Of course, we don't really know what other color translations we may be seeing (e.g. screenshot utility or github?).

@AxelG-DE and @kofa73: Do the neutral #'s I'm seeing correspond to what you're seeing? And do you know what the whitepoint is of your monitor profile? (E.g. D50, D55, D65.)

@Nilvus and @johnny-bit: Thank you for clarifying things here.

**Comment by dtorop** [comment]
It gets worse: If primaries are calculated as I suggest, and are outside of the gamut of the display space, they may not overlay to produce a neutral gray. There are probably ways around this, e.g. write a solver which comes up with reasonable primaries which combine to neutral gray (and produce a nice yellow).

I know that there's talk about how "modern" GUI color should be specified in a known colorspace which has linear gamma for color mixing. Surely someone has thought out the wrinkles of how to do this while making the color mixing result in predictable results...

I'll think this over more.

**Comment by kofa73** [comment]
@dtorop : It's a D65 profile. See here (the ZIP contains the same files): https://tech.kovacs-telekes.org/files/darktable-issue-6945-display-profile/

**Comment by AxelG-DE** [comment]
> @AxelG-DE and @kofa73: Do the neutral #'s I'm seeing correspond to what you're seeing? And do you know what the whitepoint is of your monitor profile? (E.g. D50, D55, D65.)

@dtorop I am afraid, my question sounds silly. I can imagine, with quite some efforts to replicate the numbers like 222.222.222 in e.g. gimp but how did you measure it so fast. Is there a simple and non-abusing way? I would, for now, import them into dt and use the color picker, but is this really the best way?

My Monitor is calibrated to 120cd/m^2 and 6500k is that what is abbreviated with D65?

I hope the way I am asking is not disqualifying me for future issues :-)

**Comment by kofa73** [comment]
@dtorop 
>  The image from @kofa73 the "neutral" is worse (226,204,252). Of course, we don't really know what other color translations we may be seeing (e.g. screenshot utility or github?).

It's really quite a strong tint. The screenshot utility of course grabbed whatever it found in the buffer, so in Gimp (which assumed input was sRGB) even the greys did not read as R = G = B, so I applied the monitor profile to the image in Gimp, and then converted it to sRGB. After that, I see only very small deviations (+/-1) in the areas which appear grey to my eyes. That's the version you'll find on Github.

**Comment by dtorop** [comment]
I'm thinking the "right" thing to do is to draw/blend the three color channels in a buffer which is in a known/linear colorspace. Then to convert the buffer to display colorspace and draw it to the screen. There may be performance troubles with this. I'll try it out.

This could well be the general answer to making GUI widgets with colors in a defined colorspace -- that all the compositing has to happen before converting to display colorspace and sending to output.

**Comment by kofa73** [comment]
Do we really need a colour managed histogram? For me, it has always been just a symbol; I never cared exactly which shade of blue, red or green I was clipping, only about the fact that I was clipping something. Of course, I'm not a serious user.

**Comment by dtorop** [comment]
I know it sounds absurd to have a color-managed histogram. But some sort of color management helps if we want to specify a specific red/green/blue color, and make sure that they overlay to make a neutral color. Another solution would be masking the overlay areas and drawing them separately.

I still want to finish a vectorscope, and color management also would be meaningful there, so that the vectorscope color could have a relationship to image color.

The conversation at https://discuss.pixls.us/t/wayland-color-management/10804 is relevant. It appears that this work should be happening at the compositor level, and its just an awkward stopgap to make this happen at the application level.

**Comment by dtorop** [comment]
There's a rough version of a fix at https://github.com/darktable-org/darktable/compare/master...dtorop:histogram-profiled-draw-non-lab. It still a bit buggy (the background of the graph goes darker), but I'd hope that the neutrals are now neutral.

**Comment by aurelienpierre** [comment]
Just seeing this, it seems like you are using default darktable matrices, which are D50 white, to display on D65 screen. You need one Bradford adaptation in-between.

**Comment by dtorop** [comment]
@aurelienpierre I think that's right. At one point, I had code (yes, with LittleCMS -- good for examining profiles):

```
    cmsCIEXYZ *wtpt = cmsReadTag(display_profile, cmsSigMediaWhitePointTag);
    cmsCIExyY display_wtpt;
    cmsXYZ2xyY(&display_wtpt, wtpt);
    Lab_profile = cmsCreateLab2Profile(&display_wtpt);
```

Then specified primaries in the Lab with that custom whitepoint. The idea was that the conversion from Lab to display colorspace would do the adapation. I'll check if putting that code back in will help.

I haven't followed through the math enough to know whether adapted primaries will always overlay to be neutral, regardless of display space white point. And I found it hard in Lab to find primaries which when overlaid produce a neutral white but look good, even with one white point.

My work in https://github.com/darktable-org/darktable/compare/master...dtorop:histogram-profiled-draw-non-lab was different: specify the primaries in Adobe RGB, draw the histogram/waveform in Adobe RGB to a Cairo image surface, then convert to display profile and paint the results. The nice things:

- it's easy to pick good primaries and know that they'll produce a neutral gray when overlaid
- takes advantage of darktable.color_profiles->transform_adobe_rgb_to_display infrastructure

The not nice things:

- verbose colorspace conversion code, including in iops rgbcurve and rgblevels which show background histograms
- the image surfaces are 8-bit color, so conversion uses LittleCMS -- though buffer will probably not be larger than 700x350 even for hidpi, so it's not terribly slow

**Comment by aurelienpierre** [comment]
The code you are referring to is really too complicated for what it is supposed to do. First, yes, the former code you removed converted from Lab to destination RGB with LittleCMS2, so the white point transform was indeed taken care of internally. Now, you convert from a D65 RGB space to a D65 RGB space, but using both profiles used to describe those spaces use D50 primaries.

Anyway, I think this is all too complicated. You don't need to convert the color space of the Cairo surface through LMS, it's a vector drawing, so just convert the 3 RGB primaries you use, and spare yourself slow raster encoding. Honestly, just using sRGB primaries would be fine here, we really don't want the histogram to burn CPU cycles like that.

**Comment by dtorop** [comment]
> The code you are referring to is really too complicated for what it is supposed to do. ... we really don't want the histogram to burn CPU cycles like that

Agreed!

> just using sRGB primaries would be fine here

The catch comes from the histogram being rendered as three vector drawings blended in a raster buffer. (And the waveform & RGB parade as three raster images, blended.)

How to get the sRGB primaries (or any device-indepedent primaries) such that, when blended in display space, they'll be neutral? Chromatic adaptation of a neutral in a working space will result in a neutral in display space. But take three primaries which result in a neutral gray when blended in a working space. Convert them to another colorspace (whether with a different white point or not). Even with adaptation they won't blend to be neutral. (I worked this through by hand just now using http://www.brucelindbloom.com/index.html?Calc.html to go from RGB -> XYZ -> another RGB...)

It's possible to draw the whole histogram in sRGB (or another reasonable working space) then convert it to display space and paint it to the widget. So long as the primaries in the working space blend to neutral, the primaries in display space will too. That's where I'm going with https://github.com/dtorop/darktable/tree/histogram-profiled-draw-non-lab. It does burn CPU cycles (and a bit of memory) -- I haven't yet clocked how badly. The buffer will be <= 1MB, 8-bit color.

Another way to go is to do this all on a vector level: Use Cairo to mask the red/green/blue/cyan/magenta/yellow/neutral areas of the histogram, then fill each one. I have no idea if that's more taxing on the computer, and if the code gets uglier.

Another route is to roll things back to pre-#6135, at least for the histogram, and just choosing colors in display space. To avoid a near-invisible blue channel on some displays, it could be as simple as choosing less saturated colors (in display space).

**Comment by dtorop** [comment]
The more I think it over, rolling back much of #6135 and choosing good primaries in display space seems the way to go. Ideally there's one set of primaries which are reasonable in displays ranging from, say sRGB to Rec. 2020. (It's just that 100% in one channel and 0% in the other two isn't the answer.) Worst case, there has to be a way to examine the display profile and choose decent primaries. Choosing three display space primaries which produce a neutral overlap is trivial.

I'll try this, unless anyone advocates/steers otherwise.

Then at some point talking through a color managed GUI could be interesting -- but certainly doesn't need to be implemented by hand in a particular widget.

**Comment by AxelG-DE** [comment]
I didn't expect I kick-off such big roll !sigh!

**Comment by dtorop** [comment]
> I didn't expect I kick-off such big roll !sigh!

Well, with luck it's an improvement/simplification. I think I've nearly got a simplified version nearly working in https://github.com/darktable-org/darktable/compare/master...dtorop:histogram-profiled-draw-non-lab. More in a bit...

**Comment by TurboGit** [comment]
Sorry did not follow closely all the discussions here, but I confirm on some screen the histogram color are bad! On my Adaobe-RGB screen it is ok, but on my laptop screen which calibrated with DisplayCAL I have a strong pink color.

**Comment by dtorop** [comment]
@TurboGit That makes sense -- anything profiled to D65 or so will look bad.

I'm nearly through changes -- need to smooth rough edges and check performance. It's not totally a roll-back of #6135, but should be an improvement. I'm choosing reasonable primaries in display space which should be roughly OK ranging from sRGB to Rec.2020 displays. These can be specified via CSS. I also realize that I can overlay the channels such that primaries/secondaries/neutrals all have about the same luminosity.

**Comment by Nilvus** [comment]
> @TurboGit That makes sense -- anything profiled to D65 or so will look bad.
> 
> I'm nearly through changes -- need to smooth rough edges and check performance. It's not totally a roll-back of #6135, but should be an improvement. I'm choosing reasonable primaries in display space which should be roughly OK ranging from sRGB to Rec.2020 displays. These can be specified via CSS. I also realize that I can overlay the channels such that primaries/secondaries/neutrals all have about the same luminosity.

I always like to read new things specified via CSS. Really good to read that!

**Comment by dtorop** [comment]
#6999 is an attempt to fix this. There were mysterious CI failures for macOS -- I'm not sure if the problem is the PR or some other infrastructure...

**Comment by Nilvus** [comment]
This fail is not related to your PR. The same fails is present in all recent commits since yesterday as you can see on commits list: https://github.com/darktable-org/darktable/commits/master (I have asked @TurboGit to check that).

**Comment by aurelienpierre** [comment]
> How to get the sRGB primaries (or any device-indepedent primaries) such that, when blended in display space, they'll be neutral? Chromatic adaptation of a neutral in a working space will result in a neutral in display space. But take three primaries which result in a neutral gray when blended in a working space. Convert them to another colorspace (whether with a different white point or not). Even with adaptation they won't blend to be neutral. (I worked this through by hand just now using http://www.brucelindbloom.com/index.html?Calc.html to go from RGB -> XYZ -> another RGB...)

What we draw here are scopes, not color controls, not images. To draw such scopes, we only need their geometry. Then it's only a matter of assigning a color to such geometries, but nobody told that these colors should be exactly the primaries of the color space we count the pixels in.

Cairo has an `ADD` blending mode. If you paint surfaces red (1, 0, 0), green (0, 1, 0), and blue (0, 0, 1) and overlay the surfaces with ADD blending mode, then you will get proper secondary colors and white without having to convert the color of the histogram pixel surface. You don't need to care about the display color profile. When you say RGB = (1, 0, 0), it's always relative to the display space, and since it's a synthetic image (so we don't need to ensure color consistency with an original color, it's all made up), it's ok.

**Comment by dtorop** [comment]
@aurelienpierre Yes, all sounds right. I ended up down an over-engineering path, correcting mistakes in the prior work, but making things more elaborate. PR conversations -- and testing and bug reports -- can be helpful to get things back into focus.

As of #6999, things are simpler. The code is using `ADD` for the histogram blending. Looking back, this has been true for the stable code since way back in 72a8ba54fe6db57dccd1f1e6a634108fc9c8f91f.
